chemicalx/models/gcnbmp.py

```diff
@@ -1,14 +1,257 @@
+            else:
+                hidden = layer(hidden, prev_gcn)
+                if self.short_cut and hidden.shape == layer_input.shape:
+                    hidden = hidden + layer_input
+                hiddens.append(hidden)
+                layer_input = hidden
+                prev_gcn = hiddens[-2]
+
+        if self.concat_hidden:
+            node_feature = torch.cat(hiddens, dim=-1)
+        else:
+            node_feature = hiddens[-1]
+
+        return {"node_feature": node_feature}
+
+
+class GCNBMP(Model):
+    r"""An implementation of the GCNBMP model.
 
     .. seealso:: https://github.com/AstraZeneca/chemicalx/issues/21
     """
+
+    def __init__(
+        self,
+        *,
+        molecule_channels: int = TORCHDRUG_NODE_FEATURES,
+        num_relations: int = 4,  # TODO: This default value should be set by a dataset-specific constant
+        hidden_channels: int = 16,
+        hidden_conv_layers: int = 1,
+        out_channels: int = 1,
+    ):
+        """Instantiate the GCN-BMP model.
+        :param molecule_channels: The number of node-level features.
+        :param hidden_channels: The number of hidden layer neurons in the input layer.
+        :param hidden_conv_layers: The number of hidden layers in the encoder.
+        :param out_channels: The number of output channels.
+        """
+        super(GCNBMP, self).__init__()
+
+        self.graph_convolutions = GCNBMPEncoder(
+            molecule_channels, [hidden_channels for i in range(hidden_conv_layers)], num_relations
+        )
+
+        self.attention_readout_left = AttentionPooling(molecule_channels, hidden_channels)
+        self.attention_readout_right = AttentionPooling(molecule_channels, hidden_channels)
+
+        self.final = torch.nn.Linear(hidden_channels, out_channels)
+
+    def unpack(self, batch: DrugPairBatch):
+        """Return the left and right drugs PackedGraphs."""
+        return (
+            batch.drug_molecules_left,
+            batch.drug_molecules_right,
+        )
+
+    def forward(
+        self, molecules_left: torchdrug.data.graph.PackedGraph, molecules_right: torchdrug.data.graph.PackedGraph,
+    ) -> torch.FloatTensor:
+        """
+        Run a forward pass of the GCN-BMP model.
+        Args:
+            molecules_left (torchdrug.data.graph.PackedGraph): The graph of left drug and node features.
+            molecules_right (torchdrug.data.graph.PackedGraphtorch.FloatTensor): The graph of right drug and node features.
+        Returns:
+            hidden (torch.FloatTensor): A column vector of predicted synergy scores.
+        """
+        features_left = self.graph_convolutions(molecules_left, molecules_left.data_dict["node_feature"])[
+            "node_feature"
+        ]
+
+        features_right = self.graph_convolutions(molecules_right, molecules_right.data_dict["node_feature"])[
+            "node_feature"
+        ]
+
+        features_left = self.attention_readout_left(
```

Comment:
Attention should be symmetric. The same instance is used on both sides.

Reply:
This overparametrizes the model.

Reply:
Fixed, in the same spirit note that the circular correlation operator is not symmetrical un L<->R permutation. It was managed by the authors by duplicating each pair of drugs.